File: Injector/src/main/java/com/aliucord/injector/Injector.java
/*
 * This file is part of Aliucord, an Android Discord client mod.
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.injector;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.*;
import android.util.Log;
import android.widget.Toast;

import com.discord.app.AppActivity;
import com.discord.stores.StoreClientVersion;
import com.discord.stores.StoreStream;

import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.*;
import java.lang.reflect.Method;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;

import dalvik.system.BaseDexClassLoader;
import top.canyie.pine.Pine;
import top.canyie.pine.PineConfig;
import top.canyie.pine.callback.MethodHook;

@SuppressWarnings({ "ResultOfMethodCallIgnored", "JavaReflectionMemberAccess" })
public final class Injector {
    public static final String LOG_TAG = "Injector";
    private static final String DATA_URL = "https://raw.githubusercontent.com/Aliucord/Aliucord/builds/data.json";
    private static final String DEX_URL = "https://raw.githubusercontent.com/Aliucord/Aliucord/builds/Aliucord.zip";
    private static final File BASE_DIRECTORY = new File(Environment.getExternalStorageDirectory().getAbsolutePath(), "Aliucord");

    private static MethodHook.Unhook unhook;

    public static void init() {
        PineConfig.debug = new File(BASE_DIRECTORY, ".pine_debug").exists();
        PineConfig.debuggable = new File(BASE_DIRECTORY, ".debuggable").exists();
        Log.d(LOG_TAG, "Aliucord Debuggable: " + PineConfig.debuggable);
        PineConfig.disableHiddenApiPolicy = false;
        PineConfig.disableHiddenApiPolicyForPlatformDomain = false;
        Pine.disableJitInline();
        Pine.disableProfileSaver();
        disableHiddenApiPolicy();

        try {
            Log.d(LOG_TAG, "Hooking AppActivity.onCreate...");
            unhook = Pine.hook(AppActivity.class.getDeclaredMethod("onCreate", Bundle.class), new MethodHook() {
                @Override
                public void beforeCall(Pine.CallFrame callFrame) {
                    init((AppActivity) callFrame.thisObject);
                    unhook.unhook();
                    unhook = null;
                }
            });
        } catch (Throwable th) {
            Log.e(LOG_TAG, "Failed to initialize Aliucord", th);
        }
    }

    private static void error(Context ctx, String msg, Throwable th) {
        Logger.e(msg, th);
        new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(ctx, msg, Toast.LENGTH_LONG).show());
    }

    private static void init(AppActivity appActivity) {
        Logger.d("Initializing Aliucord...");
        if (!pruneArtProfile(appActivity))
            Logger.w("Failed to prune art profile");

        try {
            var dexFile = new File(appActivity.getCodeCacheDir(), "Aliucord.zip");

            var prefs = appActivity.getSharedPreferences("aliucord", Context.MODE_PRIVATE);
            boolean useLocalDex = prefs.getBoolean("AC_from_storage", false);
            File localDex;
            if (useLocalDex && (localDex = new File(BASE_DIRECTORY, "Aliucord.zip")).exists()) {
                Logger.d("Loading dex from " + localDex.getAbsolutePath());
                try (var fis = new FileInputStream(localDex)) {
                    writeAliucordZip(fis, dexFile);
                }
            } else if (!dexFile.exists()) {
                var successRef = new AtomicBoolean(true);
                var thread = new Thread(() -> {
                    try {
                        Logger.d("Checking local Discord version...");
                        var storeClientVersionField = StoreStream.class.getDeclaredField("clientVersion");
                        storeClientVersionField.setAccessible(true);
                        var clientVersionField = StoreClientVersion.class.getDeclaredField("clientVersion");
                        clientVersionField.setAccessible(true);
                        //noinspection AccessStaticViaInstance
                        var collector = StoreStream.Companion.access$getCollector$p(StoreStream.Companion);
                        var storeClientVersion = storeClientVersionField.get(collector);
                        var version = (Integer) clientVersionField.get(storeClientVersion);
                        assert version != null;
                        Logger.d("Retrieved local Discord version: " + version);

                        Logger.d("Fetching latest Discord version...");
                        var conn = (HttpURLConnection) new URL(DATA_URL).openConnection();
                        var sb = new StringBuilder();
                        String ln;
                        try (var reader = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                            while ((ln = reader.readLine()) != null) sb.append(ln);
                        }
                        var remoteVersion = new JSONObject(new JSONTokener(sb.toString())).getInt("versionCode");
                        Logger.d("Retrieved remote Discord version: " + remoteVersion);

                        if (remoteVersion > version) {
                            error(appActivity, "Your base Discord is outdated. Please reinstall using the Installer.", null);
                            successRef.set(false);
                        } else downloadLatestAliucordDex(dexFile);
                    } catch (Throwable e) {
                        error(appActivity, "Failed to install aliucord :(", e);
                        successRef.set(false);
                    }
                });
                thread.start();
                thread.join();
                if (!successRef.get()) return;
            }

            Logger.d("Adding Aliucord to the classpath...");
            addDexToClasspath(dexFile, appActivity.getClassLoader());
            var c = Class.forName("com.aliucord.Main");
            var preInit = c.getDeclaredMethod("preInit", AppActivity.class);
            var init = c.getDeclaredMethod("init", AppActivity.class);

            Logger.d("Invoking main Aliucord entry point...");
            preInit.invoke(null, appActivity);
            init.invoke(null, appActivity);
            Logger.d("Finished initializing Aliucord");
        } catch (Throwable th) {
            error(appActivity, "Failed to initialize Aliucord :(", th);
            // Delete file so it is reinstalled the next time
            try {
                new File(appActivity.getCodeCacheDir(), "Aliucord.zip").delete();
            } catch (Throwable ignored) {}
        }
    }

    /**
     * Public so it can be manually triggered from Aliucord to update itself
     * outputFile should be new File(context.getCodeCacheDir(), "Aliucord.zip");
     */
    public static void downloadLatestAliucordDex(File outputFile) throws IOException {
        Logger.d("Downloading Aliucord.zip from " + DEX_URL + "...");
        var conn = (HttpURLConnection) new URL(DEX_URL).openConnection();
        try (var is = conn.getInputStream()) {
            writeAliucordZip(is, outputFile);
        }
        Logger.d("Finished downloading Aliucord.zip");
    }

    @SuppressLint("DiscouragedPrivateApi") // this private api seems to be stable, thanks to facebook who use it in the facebook app
    private static void addDexToClasspath(File dex, ClassLoader classLoader) throws Throwable {
        Logger.d("Adding Aliucord to the classpath...");

        // https://android.googlesource.com/platform/libcore/+/58b4e5dbb06579bec9a8fc892012093b6f4fbe20/dalvik/src/main/java/dalvik/system/BaseDexClassLoader.java#59
        var pathListField = BaseDexClassLoader.class.getDeclaredField("pathList");
        pathListField.setAccessible(true);

        var pathList = pathListField.get(classLoader);
        // Android 7: https://android.googlesource.com/platform/libcore/+/refs/heads/nougat-release/dalvik/src/main/java/dalvik/system/DexPathList.java#184
        // Current latest Master: https://android.googlesource.com/platform/libcore/+/58b4e5dbb06579bec9a8fc892012093b6f4fbe20/dalvik/src/main/java/dalvik/system/DexPathList.java#214
        assert pathList != null;
        var addDexPath = pathList.getClass().getDeclaredMethod("addDexPath", String.class, File.class);
        addDexPath.setAccessible(true);
        addDexPath.invoke(pathList, dex.getAbsolutePath(), null);

        Logger.d("Successfully added Aliucord to the classpath");
    }

    private static void writeAliucordZip(InputStream is, File outputFile) throws IOException {
        try (var fos = new FileOutputStream(outputFile)) {
            int n;
            final int sixteenKB = 16384;
            byte[] buf = new byte[sixteenKB];
            while ((n = is.read(buf)) > -1) {
                fos.write(buf, 0, n);
            }
            fos.flush();
        }
    }

    /**
     * Try to prevent method inlining by deleting the usage profile used by AOT compilation
     * https://source.android.com/devices/tech/dalvik/configure#how_art_works
     */
    private static boolean pruneArtProfile(Context ctx) {
        Logger.d("Pruning ART usage profile...");
        var profile = new File("/data/misc/profiles/cur/0/" + ctx.getPackageName() + "/primary.prof");

        if (!profile.exists()) {
            return false;
        }

        if (profile.length() > 0) {
            try {
                // Delete file contents
                new FileOutputStream(profile).close();
            } catch (Throwable ignored) {
                return false;
            }
        }
        return true;
    }

    /**
     * Disables the Android Hidden API by adding an exemption for everything ("L")
     * Works by getting reflection methods through reflection,
     * then invoking those methods to get and invoke a setter for hidden API exemptions.
     * This works because the VM thinks its internals calling the hidden method,
     * since you're invoking reflection methods using reflection.
     */
    private static void disableHiddenApiPolicy() {
        try {
            var mForName = Class.class.getDeclaredMethod("forName", String.class);
            var mGetDeclaredMethod = Class.class.getDeclaredMethod("getDeclaredMethod", String.class, Class[].class);

            var cVMRuntime = mForName.invoke(null, "dalvik.system.VMRuntime");
            var mGetRuntime = (Method) mGetDeclaredMethod.invoke(cVMRuntime, "getRuntime", null);
            Objects.requireNonNull(mGetRuntime, "Failed to get getRuntime()!");

            var mSetHiddenApiExemptions = (Method) mGetDeclaredMethod.invoke(
                cVMRuntime,
                "setHiddenApiExemptions",
                new Class[]{ String[].class }
            );
            Objects.requireNonNull(mSetHiddenApiExemptions, "Failed to get setHiddenApiExemptions()!");

            var vmRuntime = mGetRuntime.invoke(null);
            Objects.requireNonNull(vmRuntime, "Failed to get VMRuntime!");
            mSetHiddenApiExemptions.invoke(vmRuntime, (Object) new String[]{ "L" });
        } catch (Throwable t) {
            Logger.e("Failed to disable hidden api policy!", t);
        }
    }
}

File: Injector/src/main/java/com/aliucord/injector/Injector.kt
/*
 * This file is part of Aliucord, an Android Discord client mod.
 * Copyright (c) 2022 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.injector

import android.Manifest
import android.annotation.SuppressLint
import android.app.Application
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.*
import android.provider.Settings
import androidx.activity.result.contract.ActivityResultContracts
import com.discord.app.App
import com.discord.app.AppActivity
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge
import org.json.JSONObject
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.system.exitProcess

/**
 * External settings key controlling whether custom core is enabled.
 */
private const val ALIUCORD_FROM_STORAGE_KEY = "AC_from_storage"

private val applicationInitialized = AtomicBoolean(false)
private val activityInitialized = AtomicBoolean(false)

/**
 * The main entrypoint, invoked by the overridden Discord class.
 * This is invoked shortly after [App.onCreate] starts executing.
 */
internal fun init(appCtx: Application) {
    if (applicationInitialized.getAndSet(true)) return

    Logger.init()
    Logger.d("Started Aliucord Injector!")

    try {
        if (!XposedBridge.disableHiddenApiRestrictions()) {
            Logger.w("Failed to disable hidden api restrictions")
        }
        if (!XposedBridge.disableProfileSaver()) {
            Logger.w("Failed to disable profile saver")
        }

        pruneArtProfile(appCtx)
    } catch (e: Exception) {
        Logger.e("Failed to setup environment", e)
    }

    try {
        Injector(appCtx).onApplicationCreate()
    } catch (t: Throwable) {
        Logger.errorToast(appCtx, "Failed to run Aliucord Injector", t)
    }
}

private class Injector(private val appCtx: Application) {
    /**
     * Aliucord's base directory in external storage that's accessible to the user, storing settings and plugins.
     */
    private val externalBaseDir = Environment.getExternalStorageDirectory().resolve("Aliucord")

    /**
     * Aliucord's main core settings file. This is used to determine whether a custom core is enabled.
     */
    private val externalSettingsFile = externalBaseDir.resolve("settings/Aliucord.json")

    /**
     * A custom core that was built and deployed to this device, accessible to the user.
     */
    private val externalCustomCoreFile = externalBaseDir.resolve("Aliucord.zip")

    /**
     * An official Aliucord core build downloaded by Injector.
     * This is inaccessible to users and is stored in internal cache.
     */
    private val internalCoreFile = appCtx.codeCacheDir.resolve("Aliucord.zip")

    /**
     * A copy of [externalCustomCoreFile], to prevent corruption while Aliucord is already running.
     * This is inaccessible to users and stored in internal cache.
     */
    private val internalCustomCoreFile = appCtx.codeCacheDir.resolve("Aliucord.custom.zip")

    /**
     * This is invoked when [App.onCreate] is called, triggering a possible early initialization of the Aliucord core
     * if permissions have been granted and the core has already been downloaded during a prior launch.
     */
    fun onApplicationCreate() {
        if (!isPermissionsGranted()) {
            restoreCoreFlow()
            return
        }

        Logger.d("Checking custom core settings")
        val useCustomCore = isUsingCustomCore()

        // Delete old custom cores copied to code cache if they exist
        if (!useCustomCore) {
            internalCustomCoreFile.delete()
        } else {
            Logger.d("Using custom Aliucord core!")
        }

        // Copy core bundle from external storage to internal cache to prevent deletion while running
        if (useCustomCore) {
            externalCustomCoreFile.copyTo(internalCustomCoreFile, overwrite = true)
        }
        // Download new stable core
        else if (!internalCoreFile.exists()) {
            restoreCoreFlow()
            return
        }

        // Load the core
        val loadTarget = if (useCustomCore) internalCustomCoreFile else internalCoreFile
        Logger.d("Adding Aliucord core ${loadTarget.absolutePath} the classpath...")
        addDexToClasspath(
            dexFile = loadTarget,
            classLoader = appCtx.classLoader,
        )

        // Start the loaded core
        try {
            startAliucord()
        } catch (t: Throwable) {
            Logger.errorToast(appCtx, "Failed to start Aliucord!", t)

            try {
                // Delete cached files so it is redownloaded the next time
                internalCoreFile.delete()
                internalCustomCoreFile.delete()
            } catch (e: Throwable) {
                Logger.e("Failed to delete cached core files", e)
            }

            if (useCustomCore && disableCustomCore()) {
                Logger.errorToast(appCtx, "Disabled loading custom core!")
            }
        }
    }

    /**
     * The "restoration" flow of Aliucord that waits for the user to launch an activity,
     * grant permissions, and then download the core and forcefully restart the process
     * to allow for early initialization of the Aliucord core.
     */
    fun restoreCoreFlow() {
        hookActivityOnCreate { activity ->
            val permissionsResult = ::requestPermissions.takeIf { !isPermissionsGranted() }?.invoke(activity)

            Thread {
                if (permissionsResult?.get() == false)
                    return@Thread

                if (!isUsingCustomCore() && !installCore())
                    return@Thread

                restartAliucord()
            }.start()
        }
    }

    /**
     * Requests the necessary storage permissions and returns a future whether they were granted.
     * This should be called on or before [AppActivity.onCreate] before the lifecycle is RESUMED.
     */
    @SuppressLint("UseKtx")
    fun requestPermissions(activity: AppActivity): CompletableFuture<Boolean> {
        val future = CompletableFuture<Boolean>()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val userMsg = "Aliucord requires the 'All files access' permission to use its folder in Internal Storage!"
            appCtx.showToast(userMsg)

            activity.registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
                val granted = Environment.isExternalStorageManager()
                if (!granted) {
                    appCtx.showToast(userMsg)
                    Logger.e("User did not grant MANAGE_EXTERNAL_STORAGE permission!")
                }
                future.complete(granted)
            }.launch(
                Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                    .setData(Uri.parse("package:${activity.packageName}"))
            )
        } else {
            val userMsg = "Aliucord requires storage permissions to use its folder in Internal Storage!"
            appCtx.showToast(userMsg)

            activity.registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
                if (!granted) {
                    appCtx.showToast(userMsg)
                    Logger.e("User did not grant WRITE_EXTERNAL_STORAGE permission!")
                }
                future.complete(granted)
            }.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        }

        return future
    }

    /**
     * Performs the whole download flow of obtaining the latest core build.
     *
     * This first checks whether the current installation is compatible with the
     * latest core and prompts to reinstall with Manager when possible.
     */
    private fun installCore(): Boolean {
        return try {
            val data = fetchBuildData()
            Logger.d("Retrieved remote build data: $data")

            if (data.discordVersion > com.discord.BuildConfig.VERSION_CODE ||
                KotlinVersion.CURRENT < data.kotlinVersionParsed
            ) {
                // TODO: launch aliucord manager reinstall
                Logger.errorToast(appCtx, "Your base Discord is outdated. Please reinstall using Aliucord Manager.")
                false
            } else {
                downloadLatestCore(outputFile = internalCoreFile)
                true
            }
        } catch (e: Throwable) {
            Logger.errorToast(appCtx, "Failed to download latest Aliucord core!", e)
            false
        }
    }

    // -------- Starting Aliucord -------- //
    /**
     * Finds the correct Aliucord core entrypoint and runs it.
     */
    private fun startAliucord() {
        Logger.d("Obtaining Aliucord core entrypoints...")

        try {
            val c = Class.forName("com.aliucord.Main")
            val onApplicationInit = c.getDeclaredMethod("onApplicationInit", Application::class.java)

            Logger.d("Starting early Aliucord core...")
            onApplicationInit.invoke(null, appCtx)
            Logger.d("Finished early starting Aliucord")
        } catch (_: NoSuchMethodException) {
            startLegacyAliucord()
        }
    }

    /**
     * Finds the old Aliucord core entrypoints and runs it.
     * Old Aliucord core did not support early init, so we wait for Activity creation.
     */
    private fun startLegacyAliucord() {
        Logger.d("Obtaining Aliucord core legacy entrypoints...")
        val c = Class.forName("com.aliucord.Main")
        val preInit = c.getDeclaredMethod("preInit", AppActivity::class.java)
        val init = c.getDeclaredMethod("init", AppActivity::class.java)

        hookActivityOnCreate { activity ->
            Logger.d("Starting Aliucord core...")
            preInit.invoke(null, activity)
            init.invoke(null, activity)
            Logger.d("Finished starting Aliucord")
        }
    }

    // -------- Utilities -------- //

    /**
     * Checks whether external storage write permissions have been granted.
     */
    private fun isPermissionsGranted(): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Environment.isExternalStorageManager()
        } else {
            appCtx.checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
        }
    }

    /**
     * Forcefully restarts the Aliucord app process.
     */
    private fun restartAliucord(): Nothing {
        Logger.i("Force restarting Aliucord...")
        try {
            val intent = appCtx.packageManager.getLaunchIntentForPackage(appCtx.packageName)
                .let { Intent.makeRestartActivityTask(it?.component) }

            appCtx.startActivity(intent)
        } catch (e: Exception) {
            Logger.e("Failed to restart Aliucord!", e)
        } finally {
            exitProcess(0)
        }
    }

    /**
     * Checks whether using an external custom Aliucord core is possible and enabled.
     */
    private fun isUsingCustomCore(): Boolean {
        val settings = try {
            if (!externalSettingsFile.exists()) {
                Logger.d("Aliucord settings file missing, skipping custom core check...")
                return false
            }
            if (!externalCustomCoreFile.exists()) {
                Logger.d("Aliucord custom core missing, skipping custom core check...")
                return false
            }

            JSONObject(externalSettingsFile.readText())
        } catch (e: Exception) {
            // This may include permission errors
            Logger.d("Failed to read external Aliucord settings, skipping custom core check...", e)
            return false
        }

        return settings.optBoolean(ALIUCORD_FROM_STORAGE_KEY, false)
    }

    /**
     * Tries to disable the setting that enables using a custom external Aliucord core, if possible.
     * If permissions were not granted to read/write to the external settings, then this fails silently.
     */
    private fun disableCustomCore(): Boolean {
        return try {
            if (!externalSettingsFile.exists()) return true

            val settings = JSONObject(externalSettingsFile.readText())
            if (!settings.optBoolean(ALIUCORD_FROM_STORAGE_KEY, false)) {
                return true
            }

            settings.put(ALIUCORD_FROM_STORAGE_KEY, false)
            externalSettingsFile.writeText(settings.toString())
            true
        } catch (_: Exception) {
            false
        }
    }

    /**
     * Attempts to hook [AppActivity.onCreate] if it has not been invoked yet and run a [callback]
     * when it is executed. This allows for an unlimited amount of callbacks to be registered prior to
     * the activity initializing, however disallows hooking once it's initialized. As such, it should not
     * be used multiple times throughout different parts of the same initialization flow.
     */
    private fun hookActivityOnCreate(callback: (AppActivity) -> Unit) {
        Logger.d("Hooking AppActivity.onCreate")

        if (activityInitialized.get()) {
            throw IllegalStateException("Cannot hook AppActivity.onCreate when it was already invoked!")
        }

        try {
            var unhook: XC_MethodHook.Unhook? = null
            unhook = XposedBridge.hookMethod(
                AppActivity::class.java.getDeclaredMethod("onCreate", Bundle::class.java),
                object : XC_MethodHook() {
                    override fun beforeHookedMethod(param: MethodHookParam) {
                        unhook!!.unhook()
                        activityInitialized.set(true)
                        callback(param.thisObject as AppActivity)
                    }
                },
            )
        } catch (e: Exception) {
            Logger.errorToast(appCtx, "Failed to initialize Aliuhook!", e)
            throw e
        }
    }
}
